### DISCUSSION TITLE
No way to send events to parent without breaking encapsulation

### DISCUSSION THREAD

**Discussion by hedgepigdaniel** [statelyai/xstate#2961] [Ideas]
TL;DR: `sendParent` seems to be the only general way to send events from a child machine to it's parent. `sendParent` requires the child machine to know how the parent works (so that it knows what event to send). This means that if `sendParent` is used, child and parent are tightly coupled, and the child is not reusable for different parents.

## Example: fruit salad
Imagine a situation where there is some kind of complex behaviour, that can be broken down into smaller behaviours. For example, aking a fruit salad requires:
 - harvesting various fruit (in any order)
 - combining them to make a salad

### Simple/good version with `onDone`
First, let's assume the child behaviour (getting fruit) can be modeled as a simple process that starts, runs, and finishes. This is possible to do nicely in xstate by invoking child machines, and using the `onDone` handler to determine when each child machine is finished. e.g.

```typescript
const fruitSaladMakerMachine = createMachine(
  {
    id: "fruitSaladMaker",
    initial: "prepareIngredients",
    context: {
      bananasCollected: false,
      passionfruitCollected: false
    },
    states: {
      prepareIngredients: {
        invoke: [
          {
            id: "harvestBananas",
            src: harvestMachine,
            onDone: { actions: [setBananasCollected] }
          },
          {
            id: "harvestPassionfruit",
            src: harvestMachine,
            onDone: { actions: [setPassionfruitCollected] }
          }
        ]
      }
    }
  }
);
```

Some important features of this:
 - There are multiple child instances of the same machine (`harvestMachine`)
 - The parent machine needs to know which one of those two machines finished - i.e. whether it was bananas or passionfruit that just got harvested. The `onDone` handler allows this, because there is a different one for each child state machine.
 - The child machine `harvestMachine` has no knowledge of the parent machine. It exposes an API, which is simply a machine that starts, and then finishes at some time. It could be reused in any parent machine without modification.

### Complex/problematic version with `sendParent`

Let's extend this problem so that the child machine API is not so simple as `start -> finish`. Maybe the child machine emits regular progress events, or it can accept multiple requests at the same time, or it occasionally asks for a lock from the parent, etc. For a simple example, lets say that `harvestMachine` runs continuously, and can now be asked multiple times to harvest a fruit, and each time it will send another one.

```typescript
const fruitSaladMakerMachine = createMachine(
  {
    id: "fruitSaladMaker",
    initial: "prepareIngredients",
    context: {
      numBananas: 0,
      numPassionfruit: 0
    },
    states: {
      prepareIngredients: {
        invoke: [
          {
            id: "harvestBananas",
            src: harvestMachine
          },
          {
            id: "harvestPassionfruit",
            src: harvestMachine
          }
        ]
      }
    },
    on: {
      FRUIT_HARVESTED: {
        actions: ["fruitHarvested"]
      },
      REQUEST_BANANA: {
        actions: ["harvestBanana"]
      },
      REQUEST_PASSIONFRUIT: {
        actions: ["harvestPassionfruit"]
      }
    }
  },
  {
    actions: {
      fruitHarvested: addPassionfruit, // What about bananas? :(
      harvestBanana: send({ type: "REQUEST_FRUIT" }, { to: "harvestBananas" }),
      harvestPassionfruit: send(
        { type: "REQUEST_FRUIT" },
        { to: "harvestPassionfruit" }
      )
    }
  }
);

const harvestMachine = createMachine({
  id: "harvest",
  initial: "idle",
  states: {
    idle: {},
    harvesting: {
      after: {
        1000: { target: "idle", actions: send("FRUIT_HARVESTED") }
      }
    }
  },
  on: {
    FRUIT_HARVESTED: {
      actions: [sendParent("FRUIT_HARVESTED")]
    },
    REQUEST_FRUIT: {
      target: "harvesting"
    }
  }
});
```

Working example: https://codesandbox.io/s/empty-moon-yj23o?file=/src/fruitSaladMakerMachine.ts:406-1359

Unfortunately this example has a bug. The problem is that there doesn't seem to be a general way for a child machine to send an event to it's parent, without knowing how it's parent works. `onDone` is limited to one event that happens once. It's possible to use `sendParent`, (as in this example) but that has 2 problems:
 - The child machine needs to know how the parent works, and what it's purpose is inside the parent - i.e. which  parent event should happen in response to whatever happened in the child. This means that the child is no longer separate from the parent - both the parent and the child need to know about each other's inner workings.
 - As in this example, if there are multiple instances of the same machine, then there is no way to distinguish between events sent from the children. For example, this state machine always adds more passionfruit, even if it was `harvestBanana` that sent a fruit.

## Related issues:
 - The Typescript types for `sendParent` don't know what events can be sent - because it doesn't know what the parent might be. I don't see how this can be fixed unless the types for the child machine refer to the parent, which is circular.
 - https://github.com/statelyai/xstate/discussions/1620

## Idea/Suggestion
What if machines could expose external events, such that a parent machine could listen to the events `emit`ted by it's children (separately from its own events), and respond to them (separately, in the context of each child machine) as it sees fit?

```typescript
export const harvestMachine = createMachine({
  id: "harvest",
  initial: "idle",
  states: {
    idle: {},
    harvesting: {
      after: {
        1000: { target: "idle", actions: send("FRUIT_HARVESTED") }
      }
    }
  },
  on: {
    FRUIT_HARVESTED: {
      actions: [emit("GIVE_FRUIT")]
    },
    REQUEST_FRUIT: {
      target: "harvesting"
    }
  }
});

export const fruitSaladMakerMachine = createMachine(
  {
    id: "fruitSaladMaker",
    initial: "prepareIngredients",
    context: {
      numBananas: 0,
      numPassionfruit: 0
    },
    states: {
      prepareIngredients: {
        invoke: [
          {
            id: "harvestBananas",
            src: harvestMachine,
            on: { GIVE_FRUIT: { actions: [ addBanana ] } }
          },
          {
            id: "harvestPassionfruit",
            src: harvestMachine,
            on: { GIVE_FRUIT: { actions: [ addPassionfruit ] } }
          }
        ]
      }
    },
    on: {
      REQUEST_BANANA: {
        actions: ["harvestBanana"]
      },
      REQUEST_PASSIONFRUIT: {
        actions: ["harvestPassionfruit"]
      }
    }
  },
  {
    actions: {
      addPassionfruit,
      addBanana,
      harvestBanana: send({ type: "REQUEST_FRUIT" }, { to: "harvestBananas" }),
      harvestPassionfruit: send(
        { type: "REQUEST_FRUIT" },
        { to: "harvestPassionfruit" }
      )
    }
  }
);
```

To clarify, there are three sets of events here:
 - The parent machine's internal events: `REQUEST_BANANA`, `REQUEST_PASSIONFRUIT`
 - The child machine's internal events: `REQUEST_FRUIT`, `FRUIT_HARVESTED`
 - The child machine's external events (those that the parent can listen to): `GIVE_FRUIT`

The point is that the child machine never deals with events that it doesn't own.

Did I miss something? Is there any way for a parent/child to send messages in both directions, while preserving that the child is not coupled to the parent, is reusable for different parents, and can be used for multiple instances in one parent?

**Comment by theKashey** [comment]
Problem can be resolved in a _classical way_, however not closing the original questions of "matching interfaces"
- option 1: factories
```
export const createHarvestMachine = ({onHarvest}) => createMachine({
  on: {
    FRUIT_HARVESTED: {
      actions: [onHarvest]
    },
});
//...
createHarvestMachine({onHarvest:sendParent("FRUIT_HARVESTED")});
```

- option 2: adopt signals (still no information which signals sub-machine will/can send)
```ts
 invoke: [
          {
            id: "harvestBananas",
            src: adoptMachine(harvestMachine, {FRUIT_HARVESTED: BANANA_HARVESTED})
          },
          {
            id: "harvestPassionfruit",
            src: adoptMachine(harvestMachine, {FRUIT_HARVESTED: PASSIONFRUIT_HARVESTED})
          }
        ]
```

**Reply by mattpocock** [reply]
@theKashey Both of these options put a hard dependency between the parent and the child, so I agree with @hedgepigdaniel that they're not complete solutions.

**Reply by theKashey** [reply]
In both options, the decision on "how" the parent would like to use the child is on the parent. It is one way relationship from the consumer side, and while it is creating a hard dependency between these two, sound like it is a goal here.

**Comment by Andarist** [comment]
> The Typescript types for sendParent don't know what events can be sent - because it doesn't know what the parent might be. I don't see how this can be fixed unless the types for the child machine refer to the parent, which is circular.

Yeah, this is quite hard to solve - unless we ditch `sendParent` as an API and force people to "share" the parent reference explicitly. That would ensure that the child would have access to the type of the parent (which wouldn't necessarily have to be its complete type, thanks to structural typing etc the parent type could define a subset of accepted events and a "complete" parent would still be accepted). Note that this is just a roughly idea and this needs a bigger thought and a thorough design.

> Invoke machine in another machine but also standalone, with resetting 

I don't think that this issue would be fixed any time soon - if we'd expose a parent reference to a child (which is somewhat a work in progress) then a condition transition could be implemented in the child that would route to a "settled" (but not final) state or the final state. Both would mean the same for the author but the pseudo-settled state would be custom (not a builtin type of state) and would just prevent putting the child in the "done" state. At the moment I don't think this is a common enough use case to warrant implementing anything into XState to make this easier.

--- 

Going back to your original issue - ye, we recognize the problem and it's worth solving. In SCXML there is [`<finalize/>`](https://www.w3.org/TR/scxml/#finalize) that could be used for stuff like this - although it can't raise any new events and I'm unsure if `_event` is mutable so I'm actually unsure if this can be used to "massage" the event itself. Either way, this feels like a problem worth solving and I've actually thought in the past about the very same API as you have proposed: just add support for `on` within `invoke`. 

One problem with this solution is that it's somewhat invoke-specific and it feels like we should try to solve this in a way that would also cover spawned actors.

**Reply by theKashey** [reply]
Wondering can this problem be solved not statically via TypeScript, but in runtime/devtools, performing some checks matching two machines relationships, which are known in runtime.
